CTB: Improve UI validation for enum values

gu-stav opened the pull request:
### What does it do?

Builds on top of https://github.com/strapi/strapi/pull/13503, to improve the enum values validation in the admin UI. Instead of testing each value directly this PR tests the `regressed` value. This will allow more values and only fail on the ones that GraphQL can not deal with.

### Why is it needed?

It helps to prevent crashing Strapi at the bootstrap phase, after a new invalid value has been added to an enum field in the admin UI.

### How to test it?

1. Navigate to CTB
2. Add a new enum field to a content-type
3. Add invalid values (such as `--1`)
4. Validate a frontend validation error is shown

### Related issue(s)/PR(s)

- Refs https://github.com/strapi/strapi/pull/13503
- Fixes https://github.com/strapi/strapi/issues/13428

Convly reviewed (commented):
Thanks for the change. It looks like we're also operating the doesNotStartWithANumber check in the [content-type domain code](https://github.com/strapi/strapi/blob/typescript/schema-types-generation/packages/core/strapi/lib/core/domain/content-type/validator.js#L58-L62).
Should we add some consistency between both the front & back validations?

gu-stav commented:
@Convly I think it's a good idea! I've added the GraphQL enum check and another for empty values - duplicates are checked already.

I've changed the empty values check from checking the raw values to checking the regressed values, as this is the behavior I want I think.

Does that work for you?

Convly commented on packages/core/content-type-builder/admin/src/translations/en.json:
```diff
@@ -60,7 +60,7 @@
   "error.validation.enum-duplicate": "Duplicate values are not allowed",
   "error.validation.enum-empty-string": "Empty strings are not allowed",
-  "error.validation.enum-number": "Values cannot start with a number",
+  "error.validation.enum-regex": "At least one value contains invalid characters. Values should always have an alphabetical character preceding any number.",
```
Preceding the `first` number maybe?
Otherwise, it would mean that you couldn't have `a12` as `2` is preceded by a number and not an alphabetical character.
What do you think?

gu-stav replied:
Yeah, that makes sense 👍🏼

gu-stav replied:
Adressed in 1973fb57f6afd68de29492f1e21b850bfe0f8268

Convly commented on packages/core/strapi/lib/core/domain/content-type/validator.js:
```diff
@@ -24,6 +24,9 @@ const LIFECYCLES = [
 ];
 
+// See GraphQL Spec https://spec.graphql.org/June2018/#sec-Names
+const GRAPHQL_ENUM_REGEX = new RegExp('^[_A-Za-z][_0-9A-Za-z]*$');
```
Maybe it's better if we're talking about the enum regexp instead?
I would prefer we avoid coupling/referencing GraphQL to core code

gu-stav replied:
I agree, but on the other hand the GraphQL is the reason why need to this. Therefore I'd prefer to explicit here and to explain why ...

gu-stav replied:
Added in f880614ad8b69d0f6cd2be4192acea571e14bc4d Does that work for you?

Convly commented on packages/core/strapi/lib/core/domain/content-type/validator.js:
```diff
@@ -54,18 +57,25 @@ const contentTypeSchemaValidator = yup.object().shape({
+
+            // should match the GraphQL regex
+            if (!regressedValues.every(value => GRAPHQL_ENUM_REGEX.test(value))) {
+              const message = `Invalid enumervarion value. Values should always have an alphabetical character preceding any number. Update your enumeration '${attrName}'.`;
```
```suggestion
              const message = `Invalid enumeration value. Values should always have an alphabetical character preceding any number. Update your enumeration '${attrName}'.`;
```

+ same comment as above regarding the alphabetical character preceding "any" number

gu-stav replied:
Adressed in 1973fb57f6afd68de29492f1e21b850bfe0f8268

Convly reviewed (changes requested):
Thanks for the update, I just have some questions

pwizla commented:
Please let me know when this one is merged so that I can merge documentation [PR #984](https://github.com/strapi/documentation/pull/984) 😊

gu-stav commented:
@pwizla I will. Doing my best to get this one merged, so that we do not have to introduce the disclaimer :)

petersg83 commented on packages/core/content-type-builder/admin/src/translations/en.json:
```diff
@@ -60,7 +60,7 @@
   "error.contentTypeName.reserved-name": "This name cannot be used in your project as it might break other functionalities",
   "error.validation.enum-duplicate": "Duplicate values are not allowed",
```
What do you think of : `Duplicate values are not allowed (special characters are not counted)` ?

gu-stav replied:
I'd go for: "Duplicate values are not allowed (only alphanumeric characters are taken into account)"

gu-stav replied:
Done in 87435f5593e8395dc67592997c6e81efdb21bdef

petersg83 commented on packages/core/content-type-builder/admin/src/translations/en.json:
```diff
@@ -60,7 +60,7 @@
   "error.contentTypeName.reserved-name": "This name cannot be used in your project as it might break other functionalities",
   "error.validation.enum-duplicate": "Duplicate values are not allowed",
   "error.validation.enum-empty-string": "Empty strings are not allowed",
```
`Empty values are not allowed (special characters are not counted)` ?

gu-stav replied:
Discussed above.

petersg83 commented on packages/core/content-type-builder/admin/src/translations/en.json:
```diff
@@ -60,7 +60,7 @@
   "error.validation.enum-duplicate": "Duplicate values are not allowed",
   "error.validation.enum-empty-string": "Empty strings are not allowed",
-  "error.validation.enum-number": "Values cannot start with a number",
+  "error.validation.enum-regex": "At least one value contains invalid characters. Values should start with an alphabetical character preceeding the first occurence of a number.",
```
They don't contain invalid characters, it's the order that is incorrect.

```suggestion
"error.validation.enum-regex": "At least one value is invalid. Values should have at least one alphabetical character preceeding the first occurence of a number."
```
?

gu-stav replied:
I like that one! Thanks!

gu-stav commented on packages/core/strapi/lib/core/domain/content-type/validator.js:
```diff
@@ -54,22 +68,29 @@ const contentTypeSchemaValidator = yup.object().shape({
+            // should not contain empty values
+            if (regressedValues.some(value => value === '')) {
+              return this.createError({
+                message: `At least one value of the enumeration '${attrName}' appears to be empty.`,
```
What do you think about?

```suggestion
                message: `At least one value of the enumeration '${attrName}' appears to be empty. Only alphabetical characters are taken into account.`,
```

If we say ONLY these are included, we imho don't have to explain what is not.

petersg83 replied:
yes that work for me :)

gu-stav replied:
Done in 87435f5593e8395dc67592997c6e81efdb21bdef

gu-stav commented on packages/core/content-type-builder/admin/src/components/FormModal/attributes/types.js:
```diff
@@ -132,14 +131,26 @@ const types = {
         .test(alreadyUsedAttributeNames(usedAttributeNames))
         .test(isNameAllowed(reservedNames))
-        .matches(ENUM_REGEX, errorsTrads.regex)
+        .matches(GRAPHQL_ENUM_REGEX, errorsTrads.regex)
```
Oh it does - the ENUM regex was taken from the GraphQL spec, because GraphQL defines the minimum we can do with these values (every other API is more flexible). This logic should ideally live in the GraphQL plugin and not be coupled in core, but that would be a bigger refactoring. I've tried to explain that in the code comment ...

petersg83 replied:
Yes I know (I was the one that did it at that time :p), but here it is the attribute name no? Nothing to do with GraphQL

gu-stav replied:
We discussed the issue on discord: for now we keep it as it is. I've create a follow-up issue that should be adressed at some point: https://github.com/strapi/strapi/issues/13761

gu-stav commented on packages/core/strapi/lib/core/domain/content-type/validator.js:
```diff
@@ -54,22 +68,29 @@ const contentTypeSchemaValidator = yup.object().shape({
+
+            // should match the GraphQL regex
+            if (!regressedValues.every(value => GRAPHQL_ENUM_REGEX.test(value))) {
+              const message = `Invalid enumeration value. Values should start with an alphabetical character preceeding the first occurence of a number. Update your enumeration '${attrName}'.`;
```
I'd like to keep that one in sync with your other suggestion: `"Invalid enumeration value. Values should have at least one alphabetical character preceeding the first occurence of a number. Update your enumeration '${attrName}'."`

Would that work for you?

petersg83 replied:
yes :)

gu-stav replied:
Done in 87435f5593e8395dc67592997c6e81efdb21bdef

petersg83 reviewed (approved):
LGTM :)
